pkg/signatures/digester.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/gardener/component-cli/pkg/logger"
+
+	cdv2 "github.com/gardener/component-spec/bindings-go/apis/v2"
+	"github.com/gardener/component-spec/bindings-go/apis/v2/signatures"
+	cdoci "github.com/gardener/component-spec/bindings-go/oci"
+)
+
+type Digester struct {
+	ociClient       ociclient.Client
+	hasher          signatures.Hasher
+	skipAccessTypes map[string]bool
+}
+
+func NewDigester(ociClient ociclient.Client, hasher signatures.Hasher, skipAccessTypes []string) *Digester {
+	skipAccessTypesMap := map[string]bool{}
+	for _, v := range skipAccessTypes {
+		skipAccessTypesMap[v] = true
+	}
+	return &Digester{
+		ociClient:       ociClient,
+		hasher:          hasher,
+		skipAccessTypes: skipAccessTypesMap,
+	}
+
+}
+
+func (d *Digester) DigestForResource(ctx context.Context, cd cdv2.ComponentDescriptor, res cdv2.Resource) (*cdv2.DigestSpec, error) {
+	//skip ignored access type
+	if _, ok := d.skipAccessTypes[res.Access.Type]; ok {
+		return nil, nil
+	}
+
+	switch res.Access.Type {
+	case cdv2.OCIRegistryType:
+		return d.digestForOciArtifact(ctx, cd, res)
+	case cdv2.LocalOCIBlobType:
+		return d.digestForLocalOciBlob(ctx, cd, res)
+	case cdv2.S3AccessType:
+		return d.digestForS3Access(ctx, cd, res)
+	case "None":
+		logger.Log.V(5).Info(fmt.Sprintf("access type %s found in %s %s", res.Access.Type, cd.Name, cd.Version))
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("access type %s not supported", res.Access.Type)
+	}
+}
+
+func (d *Digester) digestForLocalOciBlob(ctx context.Context, componentDescriptor cdv2.ComponentDescriptor, res cdv2.Resource) (*cdv2.DigestSpec, error) {
+	if res.Access.GetType() != cdv2.LocalOCIBlobType {
+		return nil, fmt.Errorf("unsupported access type: %s", res.Access.Type)
+	}
+
+	repoctx := cdv2.OCIRegistryRepository{}
+	if err := componentDescriptor.GetEffectiveRepositoryContext().DecodeInto(&repoctx); err != nil {
+		return nil, fmt.Errorf("unable to decode repository context: %w", err)
+	}
+
+	tmpfile, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("unable to create tempfile: %w", err)
+	}
+	defer tmpfile.Close()
+
+	resolver := cdoci.NewResolver(d.ociClient)
+	_, blobResolver, err := resolver.ResolveWithBlobResolver(ctx, &repoctx, componentDescriptor.Name, componentDescriptor.Version)
+	if err != nil {
+		return nil, fmt.Errorf("unable to resolve component descriptor: %w", err)
+	}
+	if _, err := blobResolver.Resolve(ctx, res, tmpfile); err != nil {
+		return nil, fmt.Errorf("unable to resolve blob: %w", err)
+	}
+
+	if _, err := tmpfile.Seek(0, io.SeekStart); err != nil {
+		return nil, fmt.Errorf("unable to seek to beginning of tempfile: %w", err)
+	}
+	d.hasher.HashFunction.Reset()
+
+	if _, err := io.Copy(d.hasher.HashFunction, tmpfile); err != nil {
+		return nil, fmt.Errorf("unable to hash blob: %w", err)
+	}
+	return &cdv2.DigestSpec{
+		HashAlgorithm:          d.hasher.AlgorithmName,
+		NormalisationAlgorithm: string(cdv2.LocalOciBlobDigestV1),
```

Comment:
Shouldn't the normalisation algorithm be more generic, like "RawBytes" or something?

Reply:
using the GernicBlobDigestV1 as in s3 access